### ISSUE TITLE
Multiple AuthenticationHandlers not working as expected

### ISSUE THREAD

**Issue by SamWhitby** [dotnet/aspnetcore#17485]
### Describe the bug
I have a scenario where i require two different authentication handlers. One of the handlers is to be applied to all mvc requests as the default handler. The other is to be applied to only specific controllers/actions explicitly via the Authorize attribute in combination with specifying the AuthenticationSchemes parameter.

Both handlers are added in the startup like below, with the SchemeOneHandler set as the default.
``` C#
services.AddAuthentication(nameof(SchemeOneHandler))
                .AddScheme<AuthenticationSchemeOptions, SchemeOneHandler>(nameof(SchemeOneHandler), o => { })
                .AddScheme<AuthenticationSchemeOptions, SchemeTwoHandler>(nameof(SchemeTwoHandler), o => { });
```

A global filter is then added to mvc to ensure all requests are authorized as per below.
``` C#
services.AddMvc(config =>
            {
                var policy = new AuthorizationPolicyBuilder()
                         .RequireAuthenticatedUser()
                         .Build();
                config.Filters.Add(new AuthorizeFilter(policy));
            }).SetCompatibilityVersion(CompatibilityVersion.Version_2_2);
```

In my controller I have two actions one without any Authorize attribute , this is secured by the global mvc authorize policy, and a second action that has a Authorize attribute that explicitly states the AuthenticationSchemes should be SchemeTwoHandler. See code below.
``` C#
        // GET api/values
        [HttpGet]
        public ActionResult<IEnumerable<string>> Get()
        {
            return new string[] { "value1", "value2" };
        }

        // GET api/values/5
        [HttpGet("{id}")]
        [Authorize(AuthenticationSchemes = nameof(SchemeTwoHandler))]
        public ActionResult<string> Get(int id)
        {
            return "value";
        }
```
Form the docs i have read [here](https://docs.microsoft.com/en-us/aspnet/core/security/authorization/limitingidentitybyscheme?view=aspnetcore-3.0#selecting-the-scheme-with-the-authorize-attribute) this seems like it should work, and it does...partially. 

If you make a request to the action that is secured by the specific handler everything works as expected, however as soon as you make a request to the other action that is secured by the global mvc filter the first action can no longer can be accessed. 

Am I doing something wrong here, or is this a bug?

### To Reproduce
Grab the repo [here](https://github.com/SamWhitby/NetCoreAuth), and build and run. Initial request will succeed. then make a request to the route /api/values/?One and it will succeed, Then make another request to the initial route api/values/99/?Two and it will fail.

### Further technical details
.NET Core 2.2
Visual Studio 2019

**Comment by javiercn** [comment]
@SamWhitby thanks for contacting us.

@HaoK can you take a look?

**Comment by HaoK** [comment]
This was a bug in 2.1/2.2 that we have fixed in 3.0, but decided not to backport due to workarounds that should work, see https://github.com/aspnet/AspNetCore/issues/7687 for more context.

You can work around this issue by turning off
`AllowCombiningAuthorizeFilters` in mvcOptions, if that doesn't work, you can plug in a custom `IAuthorizationPolicyProvider` which just calls back into the default one which has the same effect of by passing the problematic caching code.

**Comment by SamWhitby** [comment]
@HaoK, neither of the suggested workarounds seem to resolve the issue. 
Test code for AllowCombiningAuthorizeFilters workaround can be seen [here](https://github.com/SamWhitby/NetCoreAuth/commit/0e96318eb0999653ab48190486c6cdfd34a3e123)
Test code for the IAuthorizationPolicyProvider workaround can be seen [here](https://github.com/SamWhitby/NetCoreAuth/commit/dd0fe178928c15c100f8c54ee4310436f0717af5)

**Comment by HaoK** [comment]
@SamWhitby yeah the combining logic in 2.x is problematic, I was able to work around this successfully by removing the global authorize filter and having a dedicated authorize per action:

```
        // GET api/values
        [HttpGet]
        [Authorize(AuthenticationSchemes = nameof(SchemeOneHandler))]
        public ActionResult<IEnumerable<string>> Get()
        {
            return new string[] { "value1", "value2" };
        }

        // GET api/values/5
        [HttpGet("{id}")]
        [Authorize(AuthenticationSchemes = nameof(SchemeTwoHandler))]
        public ActionResult<string> Get(int id)
        {
            return "value";
        }

```

**Comment by SamWhitby** [comment]
Thanks Hao, I was aware of that workaround however adding authorize to all routes is not a feasible option for me. I will have to hold off this till moving to 3.0.

**Comment by HaoK** [comment]
Closing this out, because this is something we have fixed in 3.0, and should be already fixed in 5.0 as well, feel free to reopen this if you run into issues still on 3.0/5.0 @SamWhitby
